Sources/CasePaths/CasePathable.swift

```diff
@@ -49,60 +49,99 @@ public protocol CasePathable {
+
+  public init<Root>(
+    get: @escaping @Sendable (Root) -> Value?,
+    set: @escaping @Sendable (inout Root, Value) -> Void
+  ) {
+    self.init(embed: nil, get: get, set: set)
+  }
+
+  fileprivate init<Root>(
+    embed: (@Sendable (Value) -> Root)?,
+    get: @escaping @Sendable (Root) -> Value?,
+    set: @escaping @Sendable (inout Root, Value) -> Void
+  ) {
+    self._embed = embed ?? { _ in Unembeddable() }
+    self._get = { @Sendable in ($0 as? Root).flatMap(get) }
+    self._set = { @Sendable in
+      var root = $0 as! Root
+      set(&root, $1)
+      $0 = root
+    }
   }
 
   public init() {
     self.init(embed: { $0 }, extract: { $0 })
   }
 
-  public init<Root>(_ keyPath: CaseKeyPath<Root, Value>) {
+  public init<Root>(_ keyPath: OptionalKeyPath<Root, Value>) {
     self = Case<Root>()[keyPath: keyPath]
   }
 
-  // #if swift(>=6)
-  //   public subscript<AppendedValue>(
-  //     dynamicMember keyPath: KeyPath<Value.AllCasePaths, AnyCasePath<Value, AppendedValue>>
-  //       & Sendable
-  //   ) -> Case<AppendedValue>
-  //   where Value: CasePathable {
-  //     Case<AppendedValue>(
-  //       embed: { embed(Value.allCasePaths[keyPath: keyPath].embed($0)) },
-  //       extract: { extract(from: $0).flatMap(Value.allCasePaths[keyPath: keyPath].extract) }
-  //     )
-  //   }
-  // #else
   public subscript<AppendedValue>(
     dynamicMember keyPath: KeyPath<Value.AllCasePaths, AnyCasePath<Value, AppendedValue>>
   ) -> Case<AppendedValue>
   where Value: CasePathable {
+    get {
+      @UncheckedSendable var keyPath = keyPath
+      return Case<AppendedValue>(
+        embed: { [$keyPath] in
+          embed(Value.allCasePaths[keyPath: $keyPath.wrappedValue].embed($0))
+        },
+        get: { [$keyPath] in
+          extract(from: $0).flatMap(Value.allCasePaths[keyPath: $keyPath.wrappedValue].extract)
+        },
+        set: { [$keyPath] in
+          set(into: &$0, Value.allCasePaths[keyPath: $keyPath.wrappedValue].embed($1))
+        }
+      )
+    }
+    set {}
+  }
+
+  @_disfavoredOverload
+  public subscript<AppendedValue>(
+    dynamicMember keyPath: WritableKeyPath<Value, AppendedValue>
+  ) -> Case<AppendedValue> {
     @UncheckedSendable var keyPath = keyPath
     return Case<AppendedValue>(
-      embed: { [$keyPath] in
-        embed(Value.allCasePaths[keyPath: $keyPath.wrappedValue].embed($0))
+      get: { [$keyPath] in
+        extract(from: $0)?[keyPath: $keyPath.wrappedValue]
       },
-      extract: { [$keyPath] in
-        extract(from: $0).flatMap(Value.allCasePaths[keyPath: $keyPath.wrappedValue].extract)
+      set: { [$keyPath] in
+        guard var value = extract(from: $0) else { return }
+        value[keyPath: $keyPath.wrappedValue] = $1
+        set(into: &$0, value)
       }
     )
   }
-  // #endif
 
   public func embed(_ value: Value) -> Any {
     self._embed(value)
   }
 
   public func extract(from root: Any) -> Value? {
-    self._extract(root)
+    self._get(root)
+  }
```

Comment:
There's a bit of a mismatch in this PR where sometimes we say "extract" and sometimes we say "get" (mostly just in the initializer for symmetry with "set"). Not sure if we want to consider something else.